### DISCUSSION TITLE
Social Aspect, Relationship, Mission, Activity, Project

### DISCUSSION THREAD

**Discussion by mokun** [mars-sim/mars-sim#308] [Q&A]
What do you guys see how important "relationship" should be in mars-sim ?

Currrently, the use of the relationship score is somewhat limited. See [Relationship wiki](https://github.com/mars-sim/mars-sim/wiki/Relationship)

Do we want to leave it this way ?

What would be the one social aspect you would like to add ?

btw, we should also consider removing the use for the phoenixt's maven package (`plexus-core`) in both `RelationshipManager` and `CreditManager`. 

`plexus-core` is an old and outdated graphing API. 

```
import com.phoenixst.plexus.DefaultGraph;
import com.phoenixst.plexus.EdgePredicate;
import com.phoenixst.plexus.EdgePredicateFactory;
import com.phoenixst.plexus.Graph;
import com.phoenixst.plexus.GraphUtils;
import com.phoenixst.plexus.NoSuchNodeException;
import com.phoenixst.plexus.Traverser;
```

If we are going to transition to save in JSON, we may as well get rid of the dependency on `plexus-core` and come up with our own in-house data structure to store those relationship scores, 

or we'll need to create a translator class for converting those graph objects into JSON and vice versa.

**Comment by Urwumpe** [comment]
I think it is very important, but I don't think it is well represented by the current data structure. It should better be a local property of every Settler. Maybe we can migrate to something Dwarf Fortress like and instead of storing a weighted graph (which is better for simulating a large number of people on a single CPU), we store how a Settler thinks about the people he heard from or interacted with. 

This way, we could also safely migrate a Settler to be processed on an other node in a Cluster, instead of having to transmit many updates on the global relationship graph. And it could put some more action into Commander mode, if the social aspect is also important for the success.

If we really need to see such a graph, we should "hire" an psychologist settler to produce this as part of his work.

**Reply by mokun** [reply]
> If we really need to see such a graph, we should "hire" an psychologist settler to produce this as part of his work.

Yea. We have yet to give the psychologist any tasks. Even setting up an interview with another settler to do a psy eval would be good.!

**Reply by mokun** [reply]
> Even setting up an interview with another settler to do a psy eval would be good.!

This is another example of me wanting to use Mission to achieve that. 

Obviously, having an one-to-one meeting with the psychologist is an indoor activity and requires no EVA.

It can even be done remotely if the person is in Settlement A while the psychologist is located in settlement B.

**Reply by mokun** [reply]
> I think it is very important, but I don't think it is well represented by the current data structure. It should better be a local property of every Settler.

Yes. It's definitely a cleaner approach. 

Just finished moving the relationship data into the Mind of a person. See [commit 1](https://github.com/mars-sim/mars-sim/commit/e383cc6e60b6105b31c3c43dd7f807362c906388) and [commit 2](https://github.com/mars-sim/mars-sim/commit/342f363e0ef98af3582d40d28794e3f85a6d0e3b)

**Reply by mokun** [reply]
> instead of having to transmit many updates on the global relationship graph.

There's no more global relationship graph from now on.

**Reply by Urwumpe** [reply]
What would be needed to allow long-distance communications with settlers in other Settlements? We should get to something like a hierarchy of communication options to be considered.

1. Talk directly, locally, in the same spot.
2. Local network radio or phone call.
3. Remote radio call
4. Visit remote location physically as a business trip.

Also, setting up such a business trip should maybe be a kind of mission itself. The settler might join others on the transport mission, but there would be a purpose, why this settler goes there and the settler will organise a transport mission again for returning home. And of course, plan to include as many activities in a short time in such a business trip,  even at the expense of recreation (some people might even cut short sleep then...*innocent whistling*)

Can we support a "Mission of missions?" Or would that be too complex? The alternative, of a separate independent mission, would be wasteful for the resources. Even a small rover could take passengers with it and you should never depart with less than 2 people onboard (also for safety)

And again, I wish for some ways of exception handling. Can we maybe "teach" settlers certain activities to take place under certain triggers and conditions? And adapt those sets of exception handlers for new conditions, like "When travelling in a rover and somebody on EVA is incapacitated, then perform the 'Recover astronaut into current rover" activity."

**Reply by mokun** [reply]
> We should get to something like a hierarchy of communication options to be considered.

Currently, we pretty much assume that if a scientist needs collaborators or assistants on a scientific study, he will find them globally. 

Other than that, there's no communication per se between settlers from different settlements. 

Also, trading or delivery mission could be considered an official contact between two settlements when negotiating the value and price of goods. 

It was done by selecting a trader from each settlement and letting the two decides what to trade.

There isn't any inter-settlement communications so far.

**Reply by mokun** [reply]
> 1. Talk directly, locally, in the same spot.
> 2. Local network radio or phone call.
> 3. Remote radio call
> 4. Visit remote location physically as a business trip.

Yes intuitively, we will need to perhaps add new parts and equipment or even the necessary infrastructure in each settlement to make the communication method 2 and 3 above possible.

Then we'll need to create some *Tasks* to make use of those equipment.

**Comment by Urwumpe** [comment]
Yeah, also we could sure add a number of "team building problems" to Mars-sim, if we could find a real psychologist to assist us there. I know that such people exist for spaceflight missions, but I doubt I could learn enough from literature to abstract this into a simulation aspect. Having somebody as collaborator, who really studied this and is really competent there, might be the best option there. 

I only really know motivation theories since it is one of my favorite topics about team and community management.

**Reply by mokun** [reply]
> "team building problems"

@bevans2000 @Urwumpe 

I have in mind creating a new class called **Activity**. It will be similar to **Mission** in the sense that it requires more than 1 person doing it. 

So your concept of Team Building Exercises can be modeled under Activity.

Initially, I suppose I would target **Activity** as indoor group activities only, such as Safety training, departmental meeting, etc. 

Do you think we we can also be able to use how scientific studies are being collaborated among researchers ?

**Reply by bevans2000** [reply]
I like the sound of it. But how different is Activity from the current Mission class? Instead of creating a new concept can we reuse the generic Mission logic with a different name for sure.
I'm working on doing some improvements to Mission in #547 to move some logic further down the class hierarchy. In theory this means a Mission could be for an internal activity. This could reuse the whole Phase and members logic.

**Reply by mokun** [reply]
> I'm working on doing some improvements to Mission in #547 to move some logic further down the class hierarchy. 

Sure. I'll wait for you to move those logic down first.

> In theory this means a Mission could be for an internal activity. This could reuse the whole Phase and members logic.

I suppose we can even have Activity as the topmost class. Under it, we would have three subclass such as Mission, Meeting and Collaboration.

Then, under Meeting, we have Training, TeamBuildingExercise and other specialized forms of meeting subclass.

**Reply by bevans2000** [reply]
We should only have new subclasses in the hierarchy if they can deliver common logic and responsibility. So would Activity have anything different to Mission>
How do you see Activity instances as being managed? We could still manage them using the Mission Manager class.

**Reply by Urwumpe** [reply]
I agree with @bevans2000 there, it makes more sense to keep things as uniform as possible, I don't think special classes would be really needed to represent different kinds of activities, unless there is REALLY a different execution behaviour. Ad-hoc or emergency activities could be such a case, most of the currently reachable cases not.

What I would suggest as alternative opinion there: Split things up. We could have a family of project-type classes or objects, that represent the static, logistic or organisational data for a mission, research project or construction project. And separate from that, but associated with this, one or more activities that coordinate the workflow and the settler actions. The project-type classes could then act also as a translator between the generic activity and the more specific entities in the data model, that are influenced by it - research results, constructions, etc. Also, ressource assignments could then be handled in such project-type classes, like having a list of "resource usage grants", that could be for all eternity or limited to a certain time window.

Also, the generic activities would be easier to implement, if they could know exception handling, like how to abort a mission or calling a "find substitute" activity for the mission commander, if one of the settlers allocated for the mission became injured before leaving.

An activity should not be using certain settlers for the execution, but roles and role assignments instead, so switching settlers is still possible and defining and instancing the activity template could be easier. The role definition could be unique for an activity or generic and reused in multiple activities (for example: Driver)

**Reply by mokun** [reply]
> So would Activity have anything different to Mission ? How do you see Activity instances as being managed? We could still manage them using the Mission Manager class.

Ok. We will not create Activity class then.

**Reply by mokun** [reply]
> We could have a family of project-type classes or objects, that represent the static, logistic or organisational data for a mission, research project or construction project.

So you suggest it's advantageous for us to create a Project class. How should it be associated with, say, Mission class ? 

> An activity should not be using certain settlers for the execution, but roles and role assignments instead, so switching settlers is still possible and defining and instancing the activity template could be easier. The role definition could be unique for an activity or generic and reused in multiple activities (for example: Driver)

I suppose you mean creating say, various JSON templates for various activities. Some kind of class will read in these templates, right ?

**Reply by Urwumpe** [reply]
> > We could have a family of project-type classes or objects, that represent the static, logistic or organisational data for a mission, research project or construction project.
> 
> So you suggest it's advantageous for us to create a Project class. How should it be associated with, say, Mission class ?

Differently:

There is a Mission, which is a special kind of "AbstractProject". This mission does the translation from the various specialized types of missions towards the activity - the goals and constraints could still be defined in special subclasses of Mission, but the internal model for the activities only knows Mission. The mission instance has a number of activities composing it, which are created from templates. The activities can refer to the mission for getting certain data - eg, the target base for travelling. The sequence, which activity is started, could be defined either as a "high-level activity" or by the old phase model, with beginning and end of a phase being used as milestones for starting the next activities or switching to a new phase after finishing all activities of that phase. 

For example, how such a sequence could look like: 

1.  In a planning phase, an activity describes how to estimate the duration of a mission, while other activities estimates the time for loading or unloading the rover, for example by executing a sub-activity for each resource that is needed for the mission.
2. In a preparation phase, an reserve rover operation describes how to select and reserve a rover for this mission and set the beginning of the loading phase, based on the initial estimates for the duration of mission and loading.
3. A special kind of action sets the beginning of the loading phase to the day the rover is available for loading.

Should be some accident the rover still be unloaded by a previous mission, which loading should begin, two escalation activities (like in Business Process Modelling Notation) could get started: One handles the case, that the rover unloading takes longer and needs a new estimate and extension of the reservation for this mission. The other handles that a new rover has to be reserved for the mission as soon as possible. 


> 
> > An activity should not be using certain settlers for the execution, but roles and role assignments instead, so switching settlers is still possible and defining and instancing the activity template could be easier. The role definition could be unique for an activity or generic and reused in multiple activities (for example: Driver)
> 
> I suppose you mean creating say, various JSON templates for various activities. Some kind of class will read in these templates, right ?

Right - the result would be an "activity template" then.

**Reply by bevans2000** [reply]
This is a good approach; makes the solution more extendable. It is similar to what was done to Task recently. We are slowly moving to consolidating the common "Activities" into the current Mission subclasses. So we have some of the parts already. I think what you are saying is the Mission is just an ordered collection of "Phases" where each phase is an activity to complete. 
To get there I think we need to complete the small refactoring to centralise common logic, e.g. new EVAMission (#547). Then we can turn the Mission structure inside out.  The specialisation would then be the MissionMeta subclasses that score the potential Mission types.

**Reply by bevans2000** [reply]
Then the ScientificStudy classes could be reworked into the same framework. In this model the ```Task``` is a Persons short term goals whilst a ```Project``` is their long-term.

**Reply by bevans2000** [reply]
I want to look into the idea of making missions workflow driven based on a number of steps, I.e. MissionStep. Missions are governed by the type hierarchy at the moment but as we discussed above we can create a set if modular steps that are linked together for execution. The existing phase logic would form the basis of different steps. We could then extend this to support the concept of Meetings and internal a activities like training and then eventually scientific study.

**Reply by Urwumpe** [reply]
That would be great - in that case, maybe "no code" people like me* could also provide new mission templates or activities by editing configuration files, re-using the existing steps.

* Currently lacking the time to fetch up with the great work you all do.

**Reply by mokun** [reply]
> making missions workflow driven based on a number of steps,

So how is the concept of a "step" different from a "phase" ?

**Reply by bevans2000** [reply]
But it would be a big change and we need a transition approach so it's not broken for all missions at once. 
I think consolidating the EVA logic proposed in #547 is a good first step. This would give up a set of logical building blocks but still is a class hierarchy. Then on the next release we start on replace the specific Mission subclasses with a generic approach one by one but based on the existing Person/Mission interactions so the new works alongside the old.

**Reply by bevans2000** [reply]
A phase at the moment is just a "tag" but a step is like a phase but with logic attached. A step would have the phase as a property so we still get the progression through a list if phases.

**Reply by bevans2000** [reply]
The no code template is a great final goal. But there are a couple of things I can't get my head round yet:

- Selecting sites to visit is specific to the goals if the mission, e,g ice mining or exploration.  That requires code
- A variable number of sites can be selected so they cannot be explicitly specified in a template 
I think there may still be a need to have explicit Mission subclasses to create the mission plan initially but the running if the mission could be more generic. The calculate next phase logic us complex at the moment relying on the super method implementation.

**Reply by mokun** [reply]
> * The calculate next phase logic us complex at the moment relying on the super method implementation.

Yes I definitely don't like those chains of super or overloaded method calls that jump from one to another, lengthening the debugging process and making whatever gone wrong very difficult to chase down.

**Reply by Urwumpe** [reply]
Well, one possible idea would be, to have a set of standard variables on a Mission or Project, which Java activity building blocks can access. 

Another option key-value pairs, which are available in the context (Mission/Project) but then we would need an exception handler activity to deal with missing data for each required value.

More complex, but more flexible solution: We give things in the simulation two traits by defining an Java interface: Some things can be settler-accessible. And if they are, they have a set of displays and controls to set certain parameters, which can be addressed by a String key. The list of parameter keys could be defined in interfaces as well, like having a DrivingControls interface for programming and as reference for implementing the template files.

In that case, an activity wouldn't need to know exact variable ID, but all that is needed is an relative address for the settler-accessible thing, for example a rover or an airlock. Some such relative addresses could maybe be passed as "subject" to activities when starting them, for example an airlock cycle could be done as small joint Activity by one or more settlers then, with the subject airlock being set by the settler starting this activity and waiting for others to join it.

For the logic how to proceed to the next phase, what about having milestones, that trigger when a number of activities leading to it have finished successfully, like it is done in commercial workflow engines? So when we define a mission step or phase, We assign each activity to two milestones: A starting milestone and an ending milestone. This way, we could also define some coarse time estimate for each milestone, by estimating for each activity how much time it roughly needs - and react to milestones slipping more than a certain threshold, for example 5 sols.

**Reply by bevans2000** [reply]
> Well, one possible idea would be, to have a set of standard variables on a Mission or Project, which Java activity building blocks can access.
> 

Certainly some standard variables will be needed.

> Another option key-value pairs, which are available in the context (Mission/Project) but then we would need an exception handler activity to deal with missing data for each required value.

Too generic and not type safe.

> 
> More complex, but more flexible solution: We give things in the simulation two traits by defining an Java interface: Some things can be settler-accessible.

There will still be a need to classify the different Missions/Projects. E.g in the dialogue we need to know which involve a Vehicle 

> 
> For the logic how to proceed to the next phase, what about having milestones, that trigger when a number of activities leading to it have finished successfully, like it is done in commercial workflow engines? So when we define a mission step or phase, We assign each activity to two milestones: A starting milestone and an ending milestone. This way, we could also define some coarse time estimate for each milestone, by estimating for each activity how much time it roughly needs - and react to milestones slipping more than a certain threshold, for example 5 sols.

Yes we could do. I was thinking more a simple linear sequence of steps as that is all we need currently. Once a step is completed it is removed from the sequence. To abort a mission, we abort the current step and replace the sequence with a travel step going home.

**Reply by bevans2000** [reply]
So I've had some thoughts on this during my morning run today.  I think it's a lot of work but I can see significant benefits is centralising around the concept of a Project. Main one is a Person can be involved in multiple projects. But we need an "evolution not revolution" approach.

I have put together this target model of where we should get finish. This includes the proposed "Activity" style projects.  (Using the cool PlantUML)
![UML Diagram of Projects](https://www.plantuml.com/plantuml/png/ZLB1Rjim3BthAtXi1xGV4FJGjUjXm0e4ox2T5ObskCcK8T5OUjctJx5Df5RZeYT1xps-HrAwpsAJ35qm1_883M7msTXzbcITW0isPzYc-1kTcFc43RGesqGnaRnSNI_n7Oe4x92bPhx5jeQUiTjZoZNyEAI4BDk3pRZWy8Xk48fyzIkcFvZEtDtD3JmJ-mtq4vBXpYYx4-nFYaywoWPoWOfaeHYpDNH9_9kIKNDhUbpZEXdxcXjZngvVYGr7GKZK7WHY0mlf1XxAuI4ou96o4BTbu9mfyYYuDOPOC3NMeHAJNy1cj9l1ulIa1jFx_gKGO5-2aMs7_lQLqKyya0kefNWKP9yl-wlsmp7cJKMByZsUrdnzaPknmjPVT9RfRz0-zKvTF_t9S-jLnUZ_XN9BREN4rmcVv-lx6xaGk2mmPhBxaQolJLOovegdOnzY6flxM9r6xHmX2pNaTZBubtevi-fXeE29syV8nMnmchC-v9iLVULC7bVUu3mtSObHMekM8hITzhCYl2_Oq8N_)

**Reply by Urwumpe** [reply]
I like that structure. Instead of a direct relation between Mind and Project, could there be a ProjectRole as relation object, which sets: 

- Which role the the Settler has in the Project
- Which settler is assigned to this role. 
- Maybe a short descriptive internal name for this role assignment for debugging (e.g. "Driver 1" and "Driver 2")?

**Reply by bevans2000** [reply]
Yes good point. I will add an Association class to the relationship. It will help identify the Leader and other Projects might have other roles

**Reply by bevans2000** [reply]
I have updated the target model and added a ProjectRole association now.
![target model](https://www.plantuml.com/plantuml/png/ZLBDRjim3BxhAOYSac4ni6lGGzUkXm7h43J3TbOask4cIuP4O-vbppx5bf1QZOkU7F7xuKU6ju6bvxuruaGQeIOqEWXv3EobOb16XW1xxtwZOf6-i8C6UU_9UUBdzMPUFo2pmHOjbyWtrvIbHsoFw4DPlk-zHylxammumt10rJCvk_xb_1_q5sprVGNRBQm-LLNyzKXMG-Ht4nhWHgmJy-C8Rg3AOpqvWoBx71WxkDbkFsHq1o6Mei6CCQOiuYNWtvGHouGP7XU-3BgknCOsvqpmWYoiOmHFpOd1rJ2Zxk0kVZGu2pXGOB9Dd3a4SdOaL4AGPVIrL9Y1oSzWFTgbOdHwI0Rf4lwICN14y6XbYxfwOVGJJwGCPYaEZ5Q7w_wP--usufMYGNvovpLlhcAf5j7oc1CjtqtEcT_KTjFrNg8NYj7_2uM6hEGpNdRud9vlR-HAmxb0YCdkXvUwD5deaHGFGsUS7yEzKvU41qLecMfI1-xrSxcSnBhhoMYopRspqQnNkS_oaAzMzDK6qhZm7rpc9XlRP3Y92XBADkTP8Dx6Mj-Q_m00)

**Comment by mokun** [comment]
Now, I have another entirely different question as follows :

Q : How would you handle a person having two meetings assigned to him ? 

In real life, I have a few meetings assigned to me from time to time on a given day to attend. 

So it's not difficult to imagine that while a person was just assigned a mission, he may also have to attend a mandatory safety briefing, etc, while his supervisor may want to see him and have a meeting with him.

So what need to be changed in the code to be able to hand multiple missions/trainings ?

**Reply by bevans2000** [reply]
Thinking of what we discussed above, each Person would have the ability to be a member of multiple Projects. There would be a priority to influence which Project a Person worked on when free.

**Reply by Urwumpe** [reply]
Also a meeting priority might be helpful. Maybe just like in Outlook, being able to tell, if a person invited to a meeting is mandatory or optional.

**Reply by mokun** [reply]
> Also a meeting priority might be helpful. Maybe just like in Outlook, being able to tell, if a person invited to a meeting is mandatory or optional.

Yes. I imagine we'll need a mission/meeting/activity priority score in TaskManager to determine which task to take. 

Also, some tasks can be done in parallel. 

e.g. If a person is very hungry/thirsty and he needs to go attend a safety meeting, he can grab his snack and water bottle and go to the conference and have the meeting while eating.  

Having said that, I'm starting to think that at some points in future, if we ever switch to simulate things at a macroscopic scale, eating and drinking should be considered background tasks, such as breathing. 

We can get rid of EatDrinkMeta and don't need to have PersonTaskManager calculate if a person should eat and drink.
